### PR TITLE
simple-auth-server added to examples

### PR DESCRIPTION
**Pull request by mygnu**
@Dowwie can you please have a look here, server implementation is complete now with tutorials

### CODE REVIEW

**Comment by fafhrd91** [comment]
I’d like to keep license same for all example. But we can add link to readme.

File: simple-auth-server/src/email_service.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
+use models::Invitation;
+use sparkpost::transmission::{
+    EmailAddress, Message, Options, Recipient, Transmission, TransmissionResponse,
+};
+
+fn get_api_key() -> String {
+    std::env::var("SPARKPOST_API_KEY").expect("SPARKPOST_API_KEY must be set")
+}
+
+#[allow(unused)]
```
Comment (DoumanAsh) [inline simple-auth-server/src/email_service.rs]: What is unused here? If you want disable such warnings, do not assign variable or use `let _`

File: simple-auth-server/src/email_service.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        .add_recipient(recipient)
+        .options(options)
+        .subject("You have been invited to join Simple-Auth-Server Rust")
+        .html(email_body);
+
+    let result = tm.send(&email);
+
+    match result {
+        Ok(res) => {
+            // println!("{:?}", &res);
```
Comment (DoumanAsh) [inline simple-auth-server/src/email_service.rs]: Remove commented out code(the same for asserts below)

File: simple-auth-server/src/register_handler.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    pub password: String,
+}
+
+// to be used to send data via the Actix actor system
+#[derive(Debug)]
+pub struct RegisterUser {
+    pub invitation_id: String,
+    pub password: String,
+}
+
```
Comment (DoumanAsh) [inline simple-auth-server/src/register_handler.rs]: Remove extra empty line, here and below

File: .travis.yml [hunk 4]
```diff
@@ -34,36 +34,37 @@ before_script:
...
-      cd template_tera && cargo check && cd ..
-      cd tls && cargo check && cd ..
-      cd rustls && cargo check && cd ..
-      cd unix-socket && cargo check && cd ..
-      cd web-cors/backend && cargo check && cd ../..
-      cd websocket && cargo check && cd ..
-      cd websocket-chat && cargo check && cd ..
-      cd websocket-chat-broker && cargo check && cd ..
-      cd websocket-tcp-chat && cargo check && cd ..
+    cd async_db && cargo check && cd ..
```
[the remark this answers is not in the document]
Comment (mygnu) [reply .travis.yml]: Done

File: simple-auth-server/README.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
+##### What?
+
+We are going to create a web-server in `rust` that only deals with user registration and authentication. I will be explaining the steps in each file as we go. The complete project code is here [repo](https://gitlab.com/mygnu/rust-auth-server/tree). Please take all this with a pinch of salt as I'm a still a noob to rust 😉.
```
[the remark this answers is not in the document]
Comment (mygnu) [reply simple-auth-server/README.md]: done

File: simple-auth-server/README.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+- [futures](https://crates.io/crates/futures) // An implementation of futures and streams featuring zero allocations, composability, and iterator-like interfaces.
+- [r2d2](https://crates.io/crates/r2d2) // A generic connection pool.
+- [serde](https://crates.io/crates/serde) // A generic serialization/deserialization framework.
+- [serde_json](https://crates.io/crates/serde_json) // A JSON serialization file format.
+- [serde_derive](https://crates.io/crates/serde_derive) // Macros 1.1 implementation of #[derive(Serialize, Deserialize)].
+- [sparkpost](https://crates.io/crates/sparkpost) // Rust bindings for sparkpost email api v1.
+- [uuid](https://crates.io/crates/uuid) // A library to generate and parse UUIDs.
+
+I have provided a brief info about the crates in use from their official description. If you want to know more about any of these crates please click on the name to go to `crates.io`.
+**Shameless plug:** `sparkpost` is my crate please leave feedback if you like/dislike it.
```
[the remark this answers is not in the document]
Comment (mygnu) [reply simple-auth-server/README.md]: done

File: simple-auth-server/README.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+Read the full tutorial series on [hgill.io](https://hgill.io)
+
+- [Auth Web Microservice with rust using Actix-Web - Complete Tutorial Part 1](https://hgill.io/posts/auth-microservice-rust-actix-web-diesel-complete-tutorial-part-1/)
+- [Auth Web Microservice with rust using Actix-Web - Complete Tutorial Part 2](https://hgill.io/posts/auth-microservice-rust-actix-web-diesel-complete-tutorial-part-2/)
+- [Auth Web Microservice with rust using Actix-Web - Complete Tutorial Part 3](https://hgill.io/posts/auth-microservice-rust-actix-web-diesel-complete-tutorial-part-3/)
+
+TODO: User Login frontend page
```
[the remark this answers is not in the document]
Comment (mygnu) [reply simple-auth-server/README.md]: removed TODO

File: simple-auth-server/src/auth_handler.rs [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+#[derive(Debug, Deserialize)]
+pub struct AuthData {
+    pub email: String,
+    pub password: String,
+}
+
+impl Message for AuthData {
+    type Result = Result<SlimUser, ServiceError>;
+}
+
```
[the remark this answers is not in the document]
Comment (mygnu) [reply simple-auth-server/src/auth_handler.rs]: done

File: simple-auth-server/src/auth_handler.rs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+impl Handler<AuthData> for DbExecutor {
+    type Result = Result<SlimUser, ServiceError>;
+    fn handle(&mut self, msg: AuthData, _: &mut Self::Context) -> Self::Result {
+        use schema::users::dsl::{users, email};
+        let conn: &PgConnection = &self.0.get().unwrap();
+        let mismatch_error = Err(ServiceError::BadRequest("Username and Password don't match".into()));
```
[the remark this answers is not in the document]
Comment (mygnu) [reply simple-auth-server/src/auth_handler.rs]: Well there are three different places where this error can be returned. I could copy/paste the same code in all three, Do you have other way in mind? can you provide an example please

File: simple-auth-server/src/auth_handler.rs [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+        let mut items = users
+            .filter(email.eq(&msg.email))
+            .load::<User>(conn)?;
+
+        if let Some(user) = items.pop() {
+            match verify(&msg.password, &user.password) {
+                Ok(matching) => {
+                    if matching { return Ok(user.into()); } else { return mismatch_error; }
```
[the remark this answers is not in the document]
Comment (mygnu) [reply simple-auth-server/src/auth_handler.rs]: done

File: simple-auth-server/src/auth_handler.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (DoumanAsh) [reply simple-auth-server/src/auth_handler.rs]: I meant it like:

```rust
        if let Some(user) = items.pop() {
            match verify(&msg.password, &user.password) {
                Ok(matching) =>  if matching {
                    return Ok(user.into());
                },
                Err(_) => (),
        }
        <your error>
```

File: simple-auth-server/src/auth_routes.rs [hunk 11]
```diff
@@ -0,0 +1,32 @@
+use actix_web::{AsyncResponder, FutureResponse, HttpResponse, HttpRequest, ResponseError, Json};
+use actix_web::middleware::identity::RequestIdentity;
+use futures::future::Future;
+use utils::create_token;
+
+use app::AppState;
+use auth_handler::{AuthData, LoggedUser};
+
+pub fn login((auth_data, req): (Json<AuthData>, HttpRequest<AppState>))
```
[the remark this answers is not in the document]
Comment (mygnu) [reply simple-auth-server/src/auth_routes.rs]: Yup, you could but I would still need HttpRequest for req.remember()

File: simple-auth-server/src/email_service.rs (hunk 1 above)
Comment (mygnu) [reply simple-auth-server/src/email_service.rs]: done

File: simple-auth-server/src/email_service.rs (hunk 2 above)
Comment (mygnu) [reply simple-auth-server/src/email_service.rs]: done

File: simple-auth-server/src/register_handler.rs (hunk 3 above)
Comment (mygnu) [reply simple-auth-server/src/register_handler.rs]: done

**Comment by fafhrd91** [comment]
My main concern is license. I’d like to keep same license for all examples, and I don’t want to see gpl especially

File: simple-auth-server/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "simple-auth-server"
+version = "0.1.0"
+license="MPL-2.0+/LGPL-3.0+"
```
Comment (DoumanAsh) [inline simple-auth-server/Cargo.toml]: Remove `license` or correct it accordingly to repo's root

**Comment by DoumanAsh** [comment]
Yep, same here, license file has been removed, so the only thing needed is to correct `*.toml` file

**Comment by DoumanAsh** [comment]
Thank you

**Comment by mygnu** [comment]
No worries

**Pull request merged**